torch/quantization/quantize_fx.py

```diff
@@ -41,20 +42,99 @@ def _fuse_fx(
+            current_module: torch.nn.Module,
+            current_module_path: str):
+        super().__init__()
+        self.prev_module_type = scope.module_type
+        self.prev_module_path = scope.module_path
+        self.scope = scope
+        self.scope.module_path = current_module_path
+        self.scope.module_type = type(current_module)
+
+    def __enter__(self):
+        return
+
+    def __exit__(self, *args):
+        self.scope.module_path = self.prev_module_path
+        self.scope.module_type = self.prev_module_type
+        return
+
+
+class QuantizationTracer(Tracer):
     def __init__(self, skipped_module_names: List[str],
                  skipped_module_classes: List[Callable]):
         super().__init__()
         self.skipped_module_names = skipped_module_names
         self.skipped_module_classes = skipped_module_classes
-
-    def is_leaf_module(self, m, module_qualified_name):
-        return (m.__module__.startswith('torch.nn') and
+        # NB: initialized the module_type of top level module to None
+        # we are assuming people won't configure the model with the type of top level
+        # module here, since people can use "" for global config
+        # We can change this if there is a use case that configures
+        # qconfig using top level module type
+        self.scope = Scope("", None)
+        self.node_name_to_scope : Dict[str, Tuple[str, Any]] = {}
+
+    def is_leaf_module(self, m: torch.nn.Module, module_qualified_name : str) -> bool:
+        return (m.__module__.startswith("torch.nn") and
                 not isinstance(m, torch.nn.Sequential)) or \
             module_qualified_name in self.skipped_module_names or \
             type(m) in self.skipped_module_classes or \
             isinstance(m, _FusedModule)
 
+    def call_module(self, m: torch.nn.Module, forward: Callable[..., Any], args : Tuple[Any, ...], kwargs : Dict[str, Any]) -> Any:
+        module_qualified_name = self.path_of_module(m)
+        if not self.is_leaf_module(m, module_qualified_name):
+            def scoped_forward(_args, _kwargs):
+                # Creating scope with information of current module
+                # scope will be restored automatically upon exit
+                with ScopeContextManager(self.scope, m, module_qualified_name):
+                    return forward(*_args, **_kwargs)
+            return scoped_forward(args, kwargs)
+        return self.create_proxy("call_module", module_qualified_name, args, kwargs)
+
+    def create_node(self, kind : str, target : Target,
+                    args : Tuple[Argument, ...], kwargs : Dict[str, Argument], name : Optional[str] = None,
+                    type_expr : Optional[Any] = None) -> Node:
+        node = super().create_node(kind, target, args, kwargs, name, type_expr)
+        if kind == "call_method":
+            self.node_name_to_scope[node.name] = (self.scope.module_path, self.scope.module_type)
+        return node
```

Comment:
do we need any test coverage that these still satisfy the contract of the parent, in case the parent behavior changes?  I'm not saying they do (I'm the wrong person to ask), probably more a question for the FX team.

Reply:
yeah James has reviewed the PR. I think there might be some concerns if the code for "call_module" changes, I can check with him again.